react/features/base/tracks/actions.js

```diff
@@ -79,19 +81,64 @@ export function createLocalTracksA(options = {}) {
         // to implement them) and the right thing to do is to ask for each
         // device separately.
         for (const device of devices) {
-            createLocalTracksF(
-                {
-                    cameraDeviceId: options.cameraDeviceId,
-                    devices: [ device ],
-                    facingMode: options.facingMode || CAMERA_FACING_MODE.USER,
-                    micDeviceId: options.micDeviceId
-                },
-                /* firePermissionPromptIsShownEvent */ false,
-                store)
-            .then(
-                localTracks => dispatch(_updateLocalTracks(localTracks)),
-                reason =>
-                    dispatch(_onCreateLocalTracksRejected(reason, device)));
+            if (getState()['features/base/tracks']
+                    .find(t => t.local && t.mediaType === device)) {
+                throw new Error(`Local track for ${device} already exists`);
+            }
+
+            const gumProcess
+                = createLocalTracksF(
+                    {
+                        cameraDeviceId: options.cameraDeviceId,
+                        devices: [ device ],
+                        facingMode:
+                            options.facingMode || CAMERA_FACING_MODE.USER,
+                        micDeviceId: options.micDeviceId
+                    },
+                    /* firePermissionPromptIsShownEvent */ false,
+                    store)
+                .then(
+                    localTracks => {
+                        // Because GUM is called for 1 device (which is actually
+                        // a media type 'audio', 'video', 'screen', etc.) we
+                        // should not get more than one JitsiTrack.
+                        if (localTracks.length !== 1) {
+                            throw new Error(
+                                `Expected exactly 1 track, but was given ${
+                                    localTracks.length} tracks for device: ${
+                                    device}.`);
+                        }
+
+                        if (gumProcess.canceled) {
+                            return _disposeTracks(localTracks)
+                                    .then(
+                                        dispatch(_trackCreateCanceled(device)));
+                        }
+
+                        return dispatch(trackAdded(localTracks[0]));
+                    },
+                    reason =>
+                        dispatch(
+                            gumProcess.canceled
+                                ? _trackCreateCanceled(device)
+                                : _onCreateLocalTracksRejected(
+                                    reason,
+                                    device)));
+
+            gumProcess.cancel = () => {
+                gumProcess.canceled = true;
```

Comment:
I think that `cancel` should be more or less a `reject` on the `Promise`. At the time that `cancel` (or `reject` in my proposal) is called, we have the following cases:

1. The `Promise` is already `resolve`d. Then `cancel`/`reject` will not do anything and cleanup will be taken care of by `_disposeAndRemoveTracks`.

2. The `Promise` is not settled yet. Then `canel`/`reject` will bring us to 3.

3. The `Promise` is `reject`ed. Then `_onCreateLocalTracksRejected` takes care of the stub in the redux store.

Then we will not need `TRACK_CREATE_CANCELED`.

Such logic will also explain easily why `TRACK_CLEATE_CANCEL` and `TRACK_CREATE_ERROR` are reduced the same way: because they're one and the same thing.

Reply:
Yes, currently those could be actually treated as the same thing, but it may be slightly confusing, because when the process is canceled we may not want to emit TRACK_CREATE_ERROR with permission denied (or any other error handling we may have in future), because we no longer care (since the action has been canceled). That was the main reason I used TRACK_CREATE_ERROR and TRACK_CREATE_CANCEL. For me it's not only about taking care of the redux state, but for the action to make sense if something else would want to listen to them.

Reply:
@lyubomir After discussion of the list I still can not really understand how to write those 3 chained promises and taking into account that the outer promise (and cancel method) have to wait for the GUM to complete. I think it will be faster if you add that on top as a separate commit.

After that I will look into fixing Promise.all problem mentioned above, because I feel like it may change, after your commit.